### PR TITLE
Apply Source Maps for Support

### PR DESCRIPTION
**Pull request by KXEinc**
## Task

[TR #XXX]()

## Why?

Why were the changes needed? What issues were the changes addressing?
(Note: some changes may seem unrelated to the ticket, this is a great place to explain further.)

We want to get better dev experience by ability to get source map natively without exposing it to non-RoleModel users in production environment. This Rack middleware serves source map only for allowed users.

## What Changed

What changed in this PR?

* [x] Add Rack middleware for serving js source maps
* [x] Configure WebPack to generate maps, append maps sources to application.js, and move maps to the different folder
* [x] Order generators by alphabet

## Screenshots

If any UI changes need to be shown off, please add screenshots here.

### CODE REVIEW

File: lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb
Comment (tingraldi) [inline lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb:26]: I like the blanket permission for `@rolemodelsoftware.com`

File: lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb
Comment (tingraldi) [inline lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb:28]: Need a blank line between methods

File: lib/generators/rolemodel/source_map/source_map/config/initializers/source_map.rb
Comment (tingraldi) [inline lib/generators/rolemodel/source_map/source_map/config/initializers/source_map.rb:5]: Grammatical error: Should be `an Array` instead of `a Array`

File: lib/generators/rolemodel/source_map/source_map_generator.rb
Comment (tingraldi) [inline lib/generators/rolemodel/source_map/source_map_generator.rb:4]: Shouldn't the class name be `SourceMapGenerator`?

File: lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js
Comment (tingraldi) [inline lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:5]: The comment should be updated since the sourceMappingURL won't be a `file:///` URL.

File: lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js
Comment (tingraldi) [inline lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: I'm fairly certain that this will need to be

```javascript
   append: "\n//# sourceMappingURL=/maps/[url]",
```

since [url] will be substituted here with the name of the map file.

**Review by tingraldi — DISMISSED** [review]
This will be very handy.

I've added some comments, but I did not give approval. I think someone that has built generators before should review this for approval.

File: lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js
Comment (KXEinc) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: It was intentional. So, if this comment contains only the file name, web-dev-tool will try to fetch the source map from the same directory as the original js bundle file. I don't want to expose `maps` folder to everyone. So, the middleware takes care of mapping URL to the `maps` folder.

File: lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb
Comment (KXEinc) [reply lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb:26]: I added this 'wild card' because our support accounts have inconsistent email prefixes. This check eliminates the manual configuration step. From a security standpoint, this is anti-patter, but most of our apps have only one user with the "@rolemodelsoftware.com" domain (sure, this is a false statement). So, this should not be a problem, but I would like to discuss it.

File: lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb
Comment (KXEinc) [reply lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb:28]: Resolved.

File: lib/generators/rolemodel/source_map/source_map/config/initializers/source_map.rb
Comment (KXEinc) [reply lib/generators/rolemodel/source_map/source_map/config/initializers/source_map.rb:5]: Resolved.

File: lib/generators/rolemodel/source_map/source_map_generator.rb
Comment (KXEinc) [reply lib/generators/rolemodel/source_map/source_map_generator.rb:4]: It should.
Resolved.

File: lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js
Comment (tingraldi) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: Nice! I didn't catch that behavior in the middleware class.

File: lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb
Comment (KXEinc) [reply lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb]: We decided not to go on this path. So, I will remove this condition.

Comment (tingraldi) [reply lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb]: On second thought, some of our deployed systems (like LightningCAD DPQ) allow users to register without validating email address ownership. So, blanket allowance for @rolemodelsoftware.com would be a problem.

**Comment by KXEinc** [comment]
Please, review this PR and share your thought with me about this feature.

File: lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js
Comment (tingraldi) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: One more thing...

The sourceMappingURL form that is in place here is the same as it would be with

```javascript
environment.config.devtool = 'source-map'
```

So, there is no need to use the `SourceMapDevToolPlugin` to achieve this outcome.

Also, one downside of using the middleware approach to guard the source maps is that this mechanism won't work for situations where we don't have a logged in user. This would be the case for the LightningCAD DPQ guest editor and may apply to other applications as well. For those situations, the source maps would have to be downloaded and manually applied.

File: lib/generators/rolemodel/all_generator.rb
Comment (Jeremy-Walton) [inline lib/generators/rolemodel/all_generator.rb:6]: What's the reasoning for changing the order of all these generators?

Comment (KXEinc) [reply lib/generators/rolemodel/all_generator.rb:6]: @Jeremy-Walton, I just like how alphabetic order looks. 
BTW, can you tell me if the running order of generators matters here?

Comment (Jeremy-Walton) [reply lib/generators/rolemodel/all_generator.rb:6]: It does. Some generators depend on other ones being run first. I.E. CSS depends on icons I think.

All that being said, RoleModel Rails is hugely out of date. The CSS does not reflect our latest learnings (doesn't pull in the design system), and it definitely does not support Rails 7

File: lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js
Comment (KXEinc) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: @tingraldi, we need to obtain a user ID to confirm that the current user can see the source map. I believe the only IP address could be used as a user ID in the 'guest' case. But the problem is most internet providers assign dynamic IP addresses. So, developers must update the 'allowed IP' list whenever new IP addresses are assigned.  It is possible to build a script for updating the 'allowed IP' list, but I don't think it is worth it (We do not have a lot of cases like the 'LightningCAD DPQ guest editor' one). So, I agree with downloading and applying manually to web-dev-tool in similar cases.

Comment (KXEinc) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: Oh, I think that I should move WebPack configs to the production environment =)

Comment (tingraldi) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js:12]: Correct!

Comment (KXEinc) [reply lib/generators/rolemodel/webpacker/templates/generated/config/webpack/development.js]: Resolved.

File: lib/generators/rolemodel/all_generator.rb
Comment (KXEinc) [reply lib/generators/rolemodel/all_generator.rb:6]: Ah, I see. Thank you for pointing me out!

Comment (KXEinc) [reply lib/generators/rolemodel/all_generator.rb]: Resolved.

File: lib/generators/rolemodel/source_map/source_map_generator.rb
Comment (Jeremy-Walton) [reply lib/generators/rolemodel/source_map/source_map_generator.rb:4]: Looks like this isn't resolved

File: lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb
Comment (Jeremy-Walton) [inline lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb:4]: Is this supposed to be inherited from a rails generator? This seems like the actual middleware, not the generator.

Comment (Jeremy-Walton) [reply lib/generators/rolemodel/source_map/source_map/app/middleware/rolemodel/source_map_generator.rb:4]: Should this be in a templates folder?

File: lib/generators/rolemodel/source_map/source_map_generator.rb
Comment (Jeremy-Walton) [inline lib/generators/rolemodel/source_map/source_map_generator.rb:8]: Should this also be copying the actual source map middleware, not just the initializer?

Comment (Jeremy-Walton) [reply lib/generators/rolemodel/source_map/source_map_generator.rb:8]: Also, the js files don't seem to be getting copied in either.

**Review by Jeremy-Walton — CHANGES_REQUESTED** [review]
Might want to try running this on a demo project to make sure it installs the files correctly. I'm not sure this is set up in a way that will actually work right now.

File: lib/generators/rolemodel/source_map/source_map_generator.rb
Comment (KXEinc) [reply lib/generators/rolemodel/source_map/source_map_generator.rb]: Resolved.

File: lib/generators/rolemodel/source_map/USAGE
Comment (Jeremy-Walton) [inline lib/generators/rolemodel/source_map/USAGE:9]: THis can be removed

**Review by Jeremy-Walton — APPROVED** [review]
Approved with the removal of the initializer comment

**Pull request merged**
